File: 0719-find-k-th-smallest-pair-distance/0719-find-k-th-smallest-pair-distance.py
class Solution(object):
    def smallestDistancePair(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        nums.sort()
        
        def helper(dist): 
            # count total num of pairs w/ diff <= dist
            left = 0
            res = 0
            for right in range(len(nums)):
                while nums[right] - nums[left] > dist:
                    left += 1             
                res += right - left
            return res
            
        left, right = 0, max(nums)
        while left < right:
            mid = left + ((right-left)//2)
            pairs = helper(mid)
            if pairs >= k:
                right = mid
            else:
                left = mid+1
        return right

File: 0719-find-k-th-smallest-pair-distance/0719-find-k-th-smallest-pair-distance.cpp
class Solution {
public:
    int smallestDistancePair(vector<int>& nums, int k) {
        std::sort(nums.begin(), nums.end());
        
        auto helper = [&](int dist) -> int {
            int left = 0, res = 0;
            for (int right = 0; right < nums.size(); ++right) {
                while (nums[right] - nums[left] > dist) {
                    ++left;
                }
                res += right -left;
            }
            return res;
        };
        
        // binary search for the kth smallest distance
        int left = 0, right = nums.back() - nums.front();
        while (left < right) {
            int mid = left+(right-left)/2;
            int pairs = helper(mid);
            if (pairs >= k) {
                right = mid;
            } else {
                left = mid+1;
            }
        }
        return left;
    }
};